pkg/endpoint/bpf.go

```diff
@@ -1146,3 +1147,62 @@ func (e *Endpoint) RequireRouting() (required bool) {
 func (e *Endpoint) RequireEndpointRoute() bool {
 	return e.DatapathConfiguration.InstallEndpointRoute
 }
+
+type netNSManager interface {
+	ConfigureNetNSForIPVLAN(netNsPath string) (mapFD, mapID int, err error)
+}
+
+// FinishIPVLANInit finishes configuring ipvlan slave device of the given endpoint.
+//
+// Unfortunately, Docker libnetwork itself moves a netdev to netns of a container
+// after the Cilium libnetwork plugin driver has responded to a `JoinEndpoint`
+// request. During the move, the netdev qdisc's get flushed by the kernel. Therefore,
+// we need to configure the ipvlan slave device in two stages.
+//
+// Because the function can be called many times for the same container in parallel,
+// we need to make the function idempotent. This is achieved by checking
+// whether the datapath map has been pinned, which indicates previous
+// successful invocation of the function for the same container, before executing
+// the configuration stages.
+//
+// FIXME: Because of the libnetwork limitation mentioned above, we cannot enforce
+// policies for an ipvlan slave before a process of a container has started. So,
+// this enables a window between the two stages during which ALL container traffic
+// is allowed.
+func (e *Endpoint) FinishIPVLANInit(mgr netNSManager, netNsPath string) error {
+
+	// No need to finish IPVLAN initialization for Docker if the endpoint isn't
+	// running with Docker.
+	if e.GetDockerNetworkID() == "" {
```

Comment:
Since we lock below, it would be more correct to do the check while holding the same lock. The current implementation races in the PR and master.
I realise that the current code has the same race, however so perhaps leave a note or something.

Reply:
`GetDockerNetworkID` acquires the Endpoint's mutex, but I agree that having one critical section per function is nice. Will access the field directly while holding the lock in the next upload.